File: benchmarks/fortran/b11_mlp.f90
! MLP training (forward + backprop, full-batch gradient descent) —
! Neural Network benchmark (MICAI). D->H->O, ReLU hidden, linear output, MSE.
! ReLU + MSE use only +,-,*,/ and max (no exp/softmax), so the result is
! bit-exact across languages. Shared 64-bit LCG (seed 42) for identical init.
! Checksum = final-epoch loss + sum of all weights (both 6 dp).
!
! Every dot-product accumulation below preserves the same operand order as the
! C reference, and the suite compiles with -ffp-contract=off so no accumulation
! is fused into an FMA (which would change the rounding versus the interpreted
! languages). Arrays are stored so the inner-loop index is contiguous; loop
! variables are multi-character to avoid aliasing the N/D/H/O/E parameters.
program mlp
  implicit none
  integer, parameter :: N = 10000, D = 16, H = 64, O = 4, E = 150
  real(8), parameter :: LR = 0.01d0

  integer(8) :: rng_state
  real(8) :: x(D, N), target(O, N)
  real(8) :: W1(D, H), b1(H), W2(H, O), b2(O)
  real(8) :: gW1(D, H), gb1(H), gW2(H, O), gb2(O)
  real(8) :: z1(H), a1(H), y(O), dy(O)
  real(8) :: scale, final_loss, epoch_loss, s, da, dz, diff, wsum
  integer :: nn, dimn, hh, oo, ee
  character(len=64) :: lbuf, wbuf

  rng_state = 42_8
  do hh = 1, H
     do dimn = 1, D
        W1(dimn, hh) = (next_double() * 2.0d0 - 1.0d0) * 0.1d0
     end do
     b1(hh) = 0.0d0
  end do
  do oo = 1, O
     do hh = 1, H
        W2(hh, oo) = (next_double() * 2.0d0 - 1.0d0) * 0.1d0
     end do
     b2(oo) = 0.0d0
  end do
  do nn = 1, N
     do dimn = 1, D
        x(dimn, nn) = next_double()
     end do
     do oo = 1, O
        target(oo, nn) = next_double()
     end do
  end do

  scale = LR / real(N, 8)
  final_loss = 0.0d0
  do ee = 1, E
     do hh = 1, H
        gb1(hh) = 0.0d0
        do dimn = 1, D
           gW1(dimn, hh) = 0.0d0
        end do
     end do
     do oo = 1, O
        gb2(oo) = 0.0d0
        do hh = 1, H
           gW2(hh, oo) = 0.0d0
        end do
     end do
     epoch_loss = 0.0d0
     do nn = 1, N
        do hh = 1, H
           s = b1(hh)
           do dimn = 1, D
              s = s + W1(dimn, hh) * x(dimn, nn)
           end do
           z1(hh) = s
           if (s > 0.0d0) then
              a1(hh) = s
           else
              a1(hh) = 0.0d0
           end if
        end do
        do oo = 1, O
           s = b2(oo)
           do hh = 1, H
              s = s + W2(hh, oo) * a1(hh)
           end do
           y(oo) = s
        end do
        do oo = 1, O
           diff = y(oo) - target(oo, nn)
           epoch_loss = epoch_loss + diff * diff
           dy(oo) = 2.0d0 * diff
        end do
        do oo = 1, O
           gb2(oo) = gb2(oo) + dy(oo)
           do hh = 1, H
              gW2(hh, oo) = gW2(hh, oo) + dy(oo) * a1(hh)
           end do
        end do
        do hh = 1, H
           da = 0.0d0
           do oo = 1, O
              da = da + W2(hh, oo) * dy(oo)
           end do
           if (z1(hh) > 0.0d0) then
              dz = da
           else
              dz = 0.0d0
           end if
           gb1(hh) = gb1(hh) + dz
           do dimn = 1, D
              gW1(dimn, hh) = gW1(dimn, hh) + dz * x(dimn, nn)
           end do
        end do
     end do
     do hh = 1, H
        b1(hh) = b1(hh) - scale * gb1(hh)
        do dimn = 1, D
           W1(dimn, hh) = W1(dimn, hh) - scale * gW1(dimn, hh)
        end do
     end do
     do oo = 1, O
        b2(oo) = b2(oo) - scale * gb2(oo)
        do hh = 1, H
           W2(hh, oo) = W2(hh, oo) - scale * gW2(hh, oo)
        end do
     end do
     final_loss = epoch_loss / real(N * O, 8)
  end do

  wsum = 0.0d0
  do hh = 1, H
     wsum = wsum + b1(hh)
     do dimn = 1, D
        wsum = wsum + W1(dimn, hh)
     end do
  end do
  do oo = 1, O
     wsum = wsum + b2(oo)
     do hh = 1, H
        wsum = wsum + W2(hh, oo)
     end do
  end do

  write(lbuf, '(F30.6)') final_loss
  write(wbuf, '(F30.6)') wsum
  write(*, '(A,1X,A)') trim(adjustl(lbuf)), trim(adjustl(wbuf))

contains

  function next_double() result(v)
    real(8) :: v
    rng_state = rng_state * 6364136223846793005_8 + 1442695040888963407_8
    v = real(ishft(rng_state, -33), 8) / real(ishft(1_8, 31), 8)
  end function next_double

end program mlp

File: benchmarks/rust/src/b11_mlp.rs
// MLP training (forward + backprop, full-batch gradient descent) —
// Neural Network benchmark (MICAI). D->H->O, ReLU hidden, linear output, MSE.
// ReLU + MSE use only +,-,*,/ and max (no exp/softmax), so the result is
// bit-exact across languages. Shared 64-bit LCG (seed 42) for identical init.
// Checksum = final-epoch loss + sum of all weights (both 6 dp).

const N: usize = 10_000;
const D: usize = 16;
const H: usize = 64;
const O: usize = 4;
const E: usize = 150;
const LR: f64 = 0.01;

fn main() {
    let mut state: u64 = 42;
    let mut next_double = || -> f64 {
        state = state
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (state >> 33) as f64 / (1u64 << 31) as f64
    };

    let mut w1 = vec![[0.0f64; D]; H];
    let mut b1 = [0.0f64; H];
    let mut w2 = vec![[0.0f64; H]; O];
    let mut b2 = [0.0f64; O];
    for h in 0..H {
        for d in 0..D {
            w1[h][d] = (next_double() * 2.0 - 1.0) * 0.1;
        }
    }
    for o in 0..O {
        for h in 0..H {
            w2[o][h] = (next_double() * 2.0 - 1.0) * 0.1;
        }
    }
    let mut x = vec![[0.0f64; D]; N];
    let mut target = vec![[0.0f64; O]; N];
    for n in 0..N {
        for d in 0..D {
            x[n][d] = next_double();
        }
        for o in 0..O {
            target[n][o] = next_double();
        }
    }

    let mut gw1 = vec![[0.0f64; D]; H];
    let mut gb1 = [0.0f64; H];
    let mut gw2 = vec![[0.0f64; H]; O];
    let mut gb2 = [0.0f64; O];
    let mut z1 = [0.0f64; H];
    let mut a1 = [0.0f64; H];
    let mut y = [0.0f64; O];
    let mut dy = [0.0f64; O];

    let scale = LR / N as f64;
    let mut final_loss = 0.0f64;
    for _ in 0..E {
        for h in 0..H {
            gb1[h] = 0.0;
            for d in 0..D {
                gw1[h][d] = 0.0;
            }
        }
        for o in 0..O {
            gb2[o] = 0.0;
            for h in 0..H {
                gw2[o][h] = 0.0;
            }
        }
        let mut epoch_loss = 0.0f64;
        for n in 0..N {
            for h in 0..H {
                let mut s = b1[h];
                for d in 0..D {
                    s += w1[h][d] * x[n][d];
                }
                z1[h] = s;
                a1[h] = if s > 0.0 { s } else { 0.0 };
            }
            for o in 0..O {
                let mut s = b2[o];
                for h in 0..H {
                    s += w2[o][h] * a1[h];
                }
                y[o] = s;
            }
            for o in 0..O {
                let diff = y[o] - target[n][o];
                epoch_loss += diff * diff;
                dy[o] = 2.0 * diff;
            }
            for o in 0..O {
                gb2[o] += dy[o];
                for h in 0..H {
                    gw2[o][h] += dy[o] * a1[h];
                }
            }
            for h in 0..H {
                let mut da = 0.0f64;
                for o in 0..O {
                    da += w2[o][h] * dy[o];
                }
                let dz = if z1[h] > 0.0 { da } else { 0.0 };
                gb1[h] += dz;
                for d in 0..D {
                    gw1[h][d] += dz * x[n][d];
                }
            }
        }
        for h in 0..H {
            b1[h] -= scale * gb1[h];
            for d in 0..D {
                w1[h][d] -= scale * gw1[h][d];
            }
        }
        for o in 0..O {
            b2[o] -= scale * gb2[o];
            for h in 0..H {
                w2[o][h] -= scale * gw2[o][h];
            }
        }
        final_loss = epoch_loss / (N * O) as f64;
    }

    let mut wsum = 0.0f64;
    for h in 0..H {
        wsum += b1[h];
        for d in 0..D {
            wsum += w1[h][d];
        }
    }
    for o in 0..O {
        wsum += b2[o];
        for h in 0..H {
            wsum += w2[o][h];
        }
    }
    println!("{:.6} {:.6}", final_loss, wsum);
}
